File: 22-generate-parentheses/generate-parentheses.java
class Solution {
    public List<String> generateParenthesis(int n) {
        List<String> ll = new ArrayList<>();
        generateParentheses(n, 0, 0, "", ll);
        return ll;
    }
    public void generateParentheses(int n, int open, int close, String ans, List<String> ll) {
        if(open == n && close == n) {
            ll.add(ans);
            return;
        }
        if(open < n) {
            generateParentheses(n, open+1, close, ans+'(', ll);
        }
        if(close < open) {
            generateParentheses(n, open, close+1, ans+')', ll);
        }
    }
}

File: 22-generate-parentheses/generate-parentheses.cpp
class Solution {
public:
    vector<string> generateParenthesis(int n) {
        vector<string> ans;
        parenthesis(0, 0, n, "", ans);
        return ans;
    }
    
    void parenthesis(int open, int close, int n, string s, vector<string>& ans) {
        if (open == n && close == n) {
            ans.push_back(s);
            return;
        }

        if (open < n) {
            parenthesis(open + 1, close, n, s + '(', ans);
        }
        
        if (close < open) {
            parenthesis(open, close + 1, n, s + ')', ans);
        }
    }
};